docs/DesignDocs/DeepCopyOutParameters.md

```diff
@@ -0,0 +1,220 @@
+  void foo[in, out, count=1] Blob *b);
+  ```
+
+  The above definition is an example of passing `b` as an `inout` parameter. Similar to the case
+  of the `in` parameter, the caller is responsible for initializing `b` and the callee can read
+  the content of the `b`. Besides, callee's modifications to `b` will now reflect on the caller
+  site with a limitation that the callee cannot re-allocate or resize `b->buf`
+  (which could lead to undefined behaviors).
+
+
+The case that remains unsupported is the `out` parameter as shown in the following example.
+```
+// EDL definition
+void foo[out, count=1] Blob *b);
+```
+The major difference between the `out` parameter and the other two is that caller is no longer responsible
+for initializing `b` (i.e., the caller can simply pass a `b` pointer with `b->buf` equals to `NULL`).
+Instead, the callee takes charge of initializing `b`, which implies that `b->buf` is variable-length.
+Such support is required by typical scenarios in which the caller cannot determine the size of `b->buf`.
+
+Proposed Design
+-----------
+
+An important constraint that affects the design of the `out` parameter support is the availability of
+shared memory between a host and an enclave. Given that this support may not be general across TEEs,
+we discuss two designs in this section.
+
+The following code snippet serves as a common example that both two design aims to support.
+- Host
+  ```
+  Blob b;
+  // b->len and b->buf have not been initialized yet.
+  foo(enclave, &b);
+  // b->len and b->buf have been initialized.
+  // b->len == 10 and b->data equals to data.
+  ```
+
+- Enclave
+  ```
+  // User-defined code
+  void foo(Blob *b)
+  {
+    b->len = 10;
+    b->buf = malloc(b->size);
+    memcpy(b->buf, data, b->size);
+  }
+  ```
+
+### Design with Shared Memory
+
+With the support of shared memory, the design is more straightforward.
+The core idea of the design is that after the callee initializes the `b`,
+it also allocates a host memory and copies the content `b` to the allocated
+memory. Note that the latter part can be done by the oeedger8r-generated code.
+
+- Host
+
+  The host invokes the oeedger8r-generated code stub as usual.
+  ```
+  // Oeedger8r-generated code
+  void foo(oe_enclave* enclave, Blob *b)
+  {
+    ...
+    oe_call_enclave_function(...);
+    ...
+  }
+  ```
+
+- Enclave
+
+  Upon the dispatching of the ECALL, the oeedger8r-generated
+  `ecall_foo` is invoked as usual. However, right after the
+  invocation of the user-defined `foo`, the design adds a piece
+  of code that allocates host memory and copies the content
+  of `b` to the allocated memory.
+
+  ```
+  // Oeedger8r-generated code
+  void ecall_foo()
+  {
+    ...
+    Blob b;
+    foo(&b);
+    ...
+    // Initialize the host memory
+    Blob *lhs = (Blob *) pargs->in;
+    lhs->len = b.len;
+    lhs->buf = oe_host_malloc(lhs->len);
+    if (lhs->buf && b.buf)
+      memcpy(lhs->buf, b.buf, lhs->len);
+    ...
+  }
+  ```
```

Comment:
Question: will the above also work for out parameters on _switchless_ calls?

Reply:
Good point. That's something we need to consider. Taking the switchless calls into account may further complicate the design without shared memory.

Reply:
switchless calls are not supported on OPTEE.